File: lintchecks/src/test/java/org/signal/lint/CardViewDetectorTest.java
package org.signal.lint;

import com.android.tools.lint.checks.infrastructure.TestFile;

import org.junit.Test;

import java.io.InputStream;
import java.util.Scanner;

import static com.android.tools.lint.checks.infrastructure.TestFiles.java;
import static com.android.tools.lint.checks.infrastructure.TestLintTask.lint;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

@SuppressWarnings("UnstableApiUsage")
public final class CardViewDetectorTest {

  private static final TestFile cardViewStub = java(readResourceAsString("CardViewStub.java"));

  @Test
  public void cardViewUsed_LogCardViewUsage_1_arg() {
    lint()
        .files(cardViewStub,
            java("package foo;\n" +
                 "import androidx.cardview.widget.CardView;\n" +
                 "public class Example {\n" +
                 "  public void buildDialog() {\n" +
                 "    new CardView(context);\n" +
                 "  }\n" +
                 "}")
        )
        .issues(AlertDialogBuilderDetector.ALERT_DIALOG_BUILDER_USAGE)
        .run()
        .expect("src/foo/Example.java:5: Warning: Using 'androidx.cardview.widget.CardView' instead of com.google.android.material.card.MaterialCardView [CardViewUsage]\n" +
                "    new CardView(context);\n" +
                "    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
        .expectFixDiffs("Fix for src/foo/Example.java line 5: Replace with new com.google.android.material.card.MaterialCardView(context):\n" +
                        "@@ -5 +5\n" +
                        "-     new CardView(context);\n" +
                        "+     new com.google.android.material.card.MaterialCardView(context);");
  }

  @Test
  public void cardViewUsed_LogCardViewUsage_2_arg() {
    lint()
        .files(cardViewStub,
            java("package foo;\n" +
                 "import androidx.cardview.widget.CardView;\n" +
                 "public class Example {\n" +
                 "  public void buildDialog() {\n" +
                 "    new CardView(context, attrs);\n" +
                 "  }\n" +
                 "}")
        )
        .issues(AlertDialogBuilderDetector.ALERT_DIALOG_BUILDER_USAGE)
        .run()
        .expect("src/foo/Example.java:5: Warning: Using 'androidx.cardview.widget.CardView' instead of com.google.android.material.card.MaterialCardView [CardViewUsage]\n" +
                "    new CardView(context, attrs);\n" +
                "    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
        .expectFixDiffs("Fix for src/foo/Example.java line 5: Replace with new com.google.android.material.card.MaterialCardView(context, attrs):\n" +
                        "@@ -5 +5\n" +
                        "-     new CardView(context, attrs);\n" +
                        "+     new com.google.android.material.card.MaterialCardView(context, attrs);");
  }

  @Test
  public void cardViewUsed_withAssignment_LogCardViewUsage_1_arg() {
    lint()
        .files(cardViewStub,
            java("package foo;\n" +
                 "import androidx.cardview.widget.CardView;\n" +
                 "public class Example {\n" +
                 "  public void buildDialog() {\n" +
                 "    CardView cardView = new CardView(context)\n" +
                 "                                         ;\n" +
                 "  }\n" +
                 "}")
        )
        .issues(AlertDialogBuilderDetector.ALERT_DIALOG_BUILDER_USAGE)
        .run()
        .expect("src/foo/Example.java:5: Warning: Using 'androidx.cardview.widget.CardView' instead of com.google.android.material.card.MaterialCardView [CardViewUsage]\n" +
                "    CardView cardView = new CardView(context)\n" +
                "                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n" +
                "0 errors, 1 warnings")
        .expectFixDiffs("Fix for src/foo/Example.java line 5: Replace with new com.google.android.material.card.MaterialCardView(context):\n" +
                        "@@ -5 +5\n" +
                        "-     CardView cardView = new CardView(context)\n" +
                        "+     CardView cardView = new com.google.android.material.card.MaterialCardView(context)");
  }

  private static String readResourceAsString(String resourceName) {
    InputStream inputStream = ClassLoader.getSystemClassLoader().getResourceAsStream(resourceName);
    assertNotNull(inputStream);
    Scanner scanner = new Scanner(inputStream).useDelimiter("\\A");
    assertTrue(scanner.hasNext());
    return scanner.next();
  }
}

File: lintchecks/src/test/java/org/signal/lint/CardViewDetectorTest.kt
package org.signal.lint

import com.android.tools.lint.checks.infrastructure.TestFiles.java
import com.android.tools.lint.checks.infrastructure.TestFiles.kotlin
import com.android.tools.lint.checks.infrastructure.TestLintTask
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Scanner

class CardViewDetectorTest {
  @Test
  fun cardViewUsed_LogCardViewUsage_1_arg() {
    TestLintTask.lint()
      .files(
        cardViewStub,
        java(
          """
          package foo;
          import androidx.cardview.widget.CardView;
          public class Example {
            public void buildCardView() {
              new CardView(context);
            }
          }
          """.trimIndent()
        )
      )
      .issues(CardViewDetector.CARD_VIEW_USAGE)
      .run()
      .expect(
        """
        src/foo/Example.java:5: Warning: Using 'androidx.cardview.widget.CardView' instead of com.google.android.material.card.MaterialCardView [CardViewUsage]
            new CardView(context);
            ~~~~~~~~~~~~~~~~~~~~~
        0 errors, 1 warnings
        """.trimIndent()
      )
      .expectFixDiffs(
        """
            Fix for src/foo/Example.java line 5: Replace with new com.google.android.material.card.MaterialCardView(context):
            @@ -5 +5
            -     new CardView(context);
            +     new com.google.android.material.card.MaterialCardView(context);
            """.trimIndent()
      )
  }

  @Test
  fun cardViewUsed_LogCardViewUsage_2_arg() {
    TestLintTask.lint()
      .files(
        cardViewStub,
        java(
          """
          package foo;
          import androidx.cardview.widget.CardView;
          public class Example {
            public void buildCardView() {
              new CardView(context, attrs);
            }
          }
          """.trimIndent()
        )
      )
      .issues(CardViewDetector.CARD_VIEW_USAGE)
      .run()
      .expect(
      """
        src/foo/Example.java:5: Warning: Using 'androidx.cardview.widget.CardView' instead of com.google.android.material.card.MaterialCardView [CardViewUsage]
            new CardView(context, attrs);
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        0 errors, 1 warnings
        """
      )
      .expectFixDiffs(
        """
        Fix for src/foo/Example.java line 5: Replace with new com.google.android.material.card.MaterialCardView(context, attrs):
        @@ -5 +5
        -     new CardView(context, attrs);
        +     new com.google.android.material.card.MaterialCardView(context, attrs);
        """.trimIndent()
      )
  }

  @Test
  fun cardViewUsed_withAssignment_LogCardViewUsage_1_arg() {
    TestLintTask.lint()
      .files(
        cardViewStub,
        java(
          """
          package foo;
          import androidx.cardview.widget.CardView;
          public class Example {
            public void buildCardView() {
              CardView cardView = new CardView(context)
                                              ;
            }
          }
          """.trimIndent()
        )
      )
      .issues(CardViewDetector.CARD_VIEW_USAGE)
      .run()
      .expect(
      """
        src/foo/Example.java:5: Warning: Using 'androidx.cardview.widget.CardView' instead of com.google.android.material.card.MaterialCardView [CardViewUsage]
            CardView cardView = new CardView(context)
                                ~~~~~~~~~~~~~~~~~~~~~
        0 errors, 1 warnings
        """.trimIndent()
      )
      .expectFixDiffs(
        """
        Fix for src/foo/Example.java line 5: Replace with new com.google.android.material.card.MaterialCardView(context):
        @@ -5 +5
        -     CardView cardView = new CardView(context)
        +     CardView cardView = new com.google.android.material.card.MaterialCardView(context)
        """.trimIndent()
      )
  }

  companion object {
    private val cardViewStub = kotlin(readResourceAsString("CardViewStub.kt"))

    private fun readResourceAsString(@Suppress("SameParameterValue") resourceName: String): String {
      val inputStream = ClassLoader.getSystemClassLoader().getResourceAsStream(resourceName)
      assertNotNull(inputStream)
      val scanner = Scanner(inputStream!!).useDelimiter("\\A")
      assertTrue(scanner.hasNext())
      return scanner.next()
    }
  }
}
